src/Common/src/Interop/OSX/Interop.CoreFoundation.CFProxy.cs

```diff
@@ -0,0 +1,145 @@
+            public IntPtr Release;
+            public IntPtr CopyDescription;
+        }
+
+        internal class CFProxy
+        {
+            private SafeCFDictionaryHandle _dictionary;
+
+            internal static readonly string kCFProxyTypeAutoConfigurationURL;
+            internal static readonly string kCFProxyTypeAutoConfigurationJavaScript;
+            internal static readonly string kCFProxyTypeFTP;
+            internal static readonly string kCFProxyTypeHTTP;
+            internal static readonly string kCFProxyTypeHTTPS;
+            internal static readonly string kCFProxyTypeSOCKS;
+
+            private static readonly IntPtr kCFProxyAutoConfigurationJavaScriptKey;
+            private static readonly IntPtr kCFProxyAutoConfigurationURLKey;
+            private static readonly IntPtr kCFProxyHostNameKey;
+            private static readonly IntPtr kCFProxyPasswordKey;
+            private static readonly IntPtr kCFProxyPortNumberKey;
+            private static readonly IntPtr kCFProxyTypeKey;
+            private static readonly IntPtr kCFProxyUsernameKey;
+
+            static CFProxy()
+            {
+                IntPtr lib = Interop.Libdl.dlopen(Interop.Libraries.CFNetworkLibrary, Interop.Libdl.RTLD_LAZY);
+                if (lib != IntPtr.Zero)
+                {
+                    kCFProxyTypeAutoConfigurationURL = LoadCFStringSymbol(lib, "kCFProxyTypeAutoConfigurationURL");
+                    kCFProxyTypeAutoConfigurationJavaScript = LoadCFStringSymbol(lib, "kCFProxyTypeAutoConfigurationJavaScript");
+                    kCFProxyTypeFTP = LoadCFStringSymbol(lib, "kCFProxyTypeFTP");
+                    kCFProxyTypeHTTP = LoadCFStringSymbol(lib, "kCFProxyTypeHTTP");
+                    kCFProxyTypeHTTPS = LoadCFStringSymbol(lib, "kCFProxyTypeHTTPS");
+                    kCFProxyTypeSOCKS = LoadCFStringSymbol(lib, "kCFProxyTypeSOCKS");
+
+                    kCFProxyAutoConfigurationJavaScriptKey = LoadSymbol(lib, "kCFProxyAutoConfigurationJavaScriptKey");
+                    kCFProxyAutoConfigurationURLKey = LoadSymbol(lib, "kCFProxyAutoConfigurationURLKey");
+                    kCFProxyHostNameKey = LoadSymbol(lib, "kCFProxyHostNameKey");
+                    kCFProxyPasswordKey = LoadSymbol(lib, "kCFProxyPasswordKey");
+                    kCFProxyPortNumberKey = LoadSymbol(lib, "kCFProxyPortNumberKey");
+                    kCFProxyTypeKey = LoadSymbol(lib, "kCFProxyTypeKey");
+                    kCFProxyUsernameKey = LoadSymbol(lib, "kCFProxyUsernameKey");
+                }
+            }
+
+            public CFProxy(SafeCFDictionaryHandle dictionary)
+            {
+                _dictionary = dictionary;
+            }
+
+            private static IntPtr LoadSymbol(IntPtr lib, string name)
+            {
+                IntPtr indirect = Interop.Libdl.dlsym(lib, name);
+                return indirect == IntPtr.Zero ? IntPtr.Zero : Marshal.ReadIntPtr(indirect);
```

Comment:
I'm starting to understand what you were implying earlier about the interop difficulties here.

You had outlined two approaches:
1. Doing everything in managed
2. Having a thick shim that exposed just a few entrypoints from managed.

Is there a 3rd option: have the shim wrap the underlying functions as close to 1:1 as possible (at least for the ones that are potentially problematic), but shielding the managed code from needing to do all of this work?

The current approach (1) might be fine, but I'm wondering if we should pursue (3) or (2) instead.

What's your opinion?

Reply:
(The response about the struct makes me feel better about (1).)

Reply:
Looks like the problematic parts are these "constants" (variables imported from system dynamic library), `CFStreamClientContext` and one delegate callback (using anything but blittable types in the declaration resulted in silent failures).

I don't see an easy way to do option (3). For the "constants" I could make a native function that translates some PAL_* enum into the k<constant name> references and then does very little work in the native code. Not sure whether it's worth it, but it would certainly look better and mismatches would result in compile-time errors instead of runtime errors. Doing anything more than that in the native code would likely result in option (2) where the whole `Uri GetProxy(Uri)` method would be implemented in native code.

Going with option (1) won't get any worse than the current state of the PR. All the nasty things are already there.

Reply:
> Going with option (1) won't get any worse than the current state of the PR

This PR is (1), right?  Just wanted to make sure I'm parsing your words correctly.

Reply:
> Doing anything more than that in the native code would likely result in option (2) where the whole Uri GetProxy(Uri) method would be implemented in native code.

Would that significantly simplify things / make them more reliable?  If so, we could put that one export into System.Net.Http.Native.dylib, compiled in only for macOS.

Reply:
> This PR is (1), right?

Yes

> Would that significantly simplify things / make them more reliable?

Possibly. It would be more resilient to any changes Apple decides to make in the API.

The Mono code was around for roughly 7 years. In that time it was revised three times with some significant changes. Two of them were due to incorrect API signature or incorrect usage of the bindings that would likely be avoided by using native code. Third one was due to incorrectly used API that resulted in crashes with multi-threaded code. The last one would likely happen even if written in native code.

> I still lean towards the managed implementation, but I don't want to be blind to the costs.

Fine with me. I just wanted to point out the implications and possible issues. The `System.Security.*` code on macOS took the direction of pushing more code into the native library, so I wanted to make sure what is the preferred approach.